jax/interpreters/partial_eval.py

```diff
@@ -1581,7 +1613,10 @@ def _const_folding_and_forwarding(jaxpr, constvals):
     # always apply invar substitutions
     eqn = eqn.replace(invars=[var_subs.get(v, v) for v in eqn.invars])
     # if any inputs are constants and we have a constant-folding rule, apply it
-    if eqn.primitive in const_fold_rules and any(v in consts for v in eqn.invars):
+    has_input_effect = any(isinstance(eff, effects.JaxprInputEffect)
+                           for eff in eqn.effects)
+    if (eqn.primitive in const_fold_rules and any(v in consts for v in eqn.invars)
+        and not has_input_effect):
```

Comment:
This could be a little conservative, hypothetically, e.g. what if we had a constant folding rule for jit (which just constant folded inside it) and we had an eqn corresponding to an application of `f` to a constant where

```python
@jit
def f(c1):
  c2 = lax.convert_element_type(c1, jnp.dtype('float32'))
  x_ref[()] = 5.
  return c2
```

This logic would block constant folding of it.

(Sharad observed: we could rule out issues here if we implemented such transformations as make-jaxpr-of-jaxpr-interpreter.)